tests are failing for your DataParallel change.
See 2.7: https://build.pytorch.org/job/pytorch-pull-requests-py2/606/console
See 3.5: https://build.pytorch.org/job/pytorch-pull-requests-py3/603/console

Same failure in all the tests:
```
Traceback (most recent call last):
  File "test_nn.py", line 995, in test_data_parallel
    out = dp.data_parallel(l, i, (0, 1))
  File "/home/jenkins/miniconda/envs/py2k/lib/python2.7/site-packages/torch/nn/parallel/data_parallel.py", line 104, in data_parallel
    replicas = replicate(module, device_ids[:len(inputs)])
  File "/home/jenkins/miniconda/envs/py2k/lib/python2.7/site-packages/torch/nn/parallel/replicate.py", line 19, in replicate
    buffer_copies = comm.broadcast_coalesced(buffers, devices)
  File "/home/jenkins/miniconda/envs/py2k/lib/python2.7/site-packages/torch/cuda/comm.py", line 56, in broadcast_coalesced
    for chunk in _take_tensors(tensors, buffer_size):
  File "/home/jenkins/miniconda/envs/py2k/lib/python2.7/site-packages/torch/cuda/comm.py", line 239, in _take_tensors
    last_type = type(tensors[0])
IndexError: list index out of range
```

--- a/test/test_cuda.py
+++ b/test/test_cuda.py
@@ -463,6 +463,32 @@ class TestCuda(TestCase):
         self._test_broadcast(torch.randn(5, 5))
 
     @unittest.skipIf(torch.cuda.device_count() < 2, "only one GPU detected")
+    def test_broadcast_coalesced(self):
+        numel = 5
+        num_bytes = numel * 8
+        tensors = [
+            torch.randn(numel).long().cuda(),
+            torch.randn(numel).cuda(),
+            torch.randn(numel).long().cuda(),
+            torch.randn(numel).long().cuda(),
+            torch.randn(numel * 2).int().cuda(),  # int is 2x shorter
+            torch.randn(numel).cuda(),
+        ]
+
+        b_tensors = [comm.broadcast(t, (0, 1)) for t in tensors]
+        for (_, bt), t in zip(b_tensors, tensors):
+            self.assertEqual(bt.get_device(), 1)
+            self.assertEqual(bt, t)
+            self.assertIsInstance(bt, type(t))
+
+        bc_tensors = comm.broadcast_coalesced(tensors, (0, 1), buffer_size=num_bytes * 5 // 2)
+        bc_tensors_t = list(zip(*bc_tensors))
+        self.assertEqual(b_tensors, bc_tensors_t)
+        for (_, bt), (_, bct) in zip(b_tensors, bc_tensors_t):
+            self.assertEqual(bt.get_device(), bct.get_device())
+            self.assertIsInstance(bct, type(bt))
+
+    @unittest.skipIf(torch.cuda.device_count() < 2, "only one GPU detected")
     def test_reduce_add(self):
         x = torch.randn(5, 5)
         y = torch.randn(5, 5)
@@ -471,6 +497,32 @@ class TestCuda(TestCase):
         result = comm.reduce_add((x_cuda, y_cuda))
         self.assertEqual(result.get_device(), 0)
         self.assertEqual(result.cpu(), x + y)
+
+    @unittest.skipIf(torch.cuda.device_count() < 2, "only one GPU detected")
+    def test_reduce_add_coalesced(self):
+        numel = 5
+        num_bytes = numel * 8
+        tensors = [
+            torch.randn(numel).long().cuda(),
+            torch.randn(numel).cuda(),
+            torch.randn(numel).long().cuda(),
+            torch.randn(numel).long().cuda(),
+            torch.randn(numel * 2).int().cuda(),  # int is 2x shorter
+            torch.randn(numel).cuda(),
+        ]
+        dup_tensors = [tensors, list(map(lambda t: t.cuda(1), tensors))]
+
+        r_tensors = list(map(comm.reduce_add, zip(*dup_tensors)))
+        for r, t in zip(r_tensors, tensors):
+            self.assertEqual(r.get_device(), t.get_device())
+            self.assertEqual(r, t * 2)
+            self.assertIsInstance(r, type(t))
+
+        rc_tensors = comm.reduce_add_coalesced(dup_tensors, buffer_size=num_bytes * 5 // 2)
+        self.assertEqual(r_tensors, rc_tensors)
+        for r, rc in zip(r_tensors, rc_tensors):
+            self.assertEqual(rc.get_device(), r.get_device())
+            self.assertIsInstance(rc, type(r))
 
     def _test_scatter(self, input, chunk_sizes=None, dim=0):
         if torch.cuda.device_count() < 2:

--- a/torch/cuda/comm.py
+++ b/torch/cuda/comm.py
@@ -108,7 +108,8 @@ def reduce_add_coalesced(inputs, destination=None, buffer_size=10485760):
     of synchronizations.
 
     Arguments:
-        inputs (Iterable[Tensor]): an iterable of tensors to add.
+        inputs (Iterable[Iterable[Tensor]]): iterable of iterables that
+            contain tensors from a single device.
         destination (int, optional): a device on which the output will be
             placed (default: current device).
         buffer_size (int): maximum size of the buffer used for coalescing
@@ -235,10 +236,13 @@ def _take_tensors(tensors, size_limit):
     """Groups tensors into lists of up to size_limit bytes"""
     buf = []
     size = 0
+    last_type = type(tensors[0]) if len(tensors) > 0 else None
     for tensor in tensors:
+        t = type(tensor)
         param_size = tensor.numel() * tensor.element_size()
-        if size + param_size > size_limit and size > 0:
+        if t is not last_type or (size + param_size > size_limit and size > 0):
             yield buf
+            last_type = t
             size = 0
             buf = []
         buf.append(tensor)

